// babel-core/lib/config/plugin.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

var Plugin = function Plugin(plugin, options, key) {
  if (plugin.name != null && typeof plugin.name !== "string") {
    throw new Error("Plugin .name must be a string, null, or undefined");
  }

  if (plugin.manipulateOptions != null && typeof plugin.manipulateOptions !== "function") {
    throw new Error("Plugin .manipulateOptions must be a function, null, or undefined");
  }

  if (plugin.post != null && typeof plugin.post !== "function") {
    throw new Error("Plugin .post must be a function, null, or undefined");
  }

  if (plugin.pre != null && typeof plugin.pre !== "function") {
    throw new Error("Plugin .pre must be a function, null, or undefined");
  }

  if (plugin.visitor != null && _typeof(plugin.visitor) !== "object") {
    throw new Error("Plugin .visitor must be an object, null, or undefined");
  }

  this.key = plugin.name || key;
  this.manipulateOptions = plugin.manipulateOptions;
  this.post = plugin.post;
  this.pre = plugin.pre;
  this.visitor = plugin.visitor;
  this.options = options;
};

exports.default = Plugin;
